File: src/day3/python/solution.py
from typing import List

def part1(pattern : List[str], slope : tuple) -> int:
    trees = 0
    down, right = 0, 0
    # I would argue the iterative "raw-loop" here is better than using a functional approach
    while (down < len(pattern)):
        # note that the pattern extends to the right infinitely, meaning the cols
        # of each row extend following modular arithmetic col % N, where N is # of cols
        if pattern[down][right % len(pattern[down])] == "#":
            trees += 1
        down += slope[0]
        right += slope[1]
    return trees

def part2(pattern : List[str], slopes : List[tuple]) -> int:
    import math
    # https://www.python.org/dev/peps/pep-0289/
    return math.prod(part1(pattern, slope) for slope in slopes) # generator exression

if __name__ == "__main__":
    pattern = open("../input.txt").read().splitlines() # M x N grid
    slopes = [(1,1), (1,3), (1,5), (1,7), (2,1)] # convention is rows down, columns right
    print("Part 1:", part1(pattern, slopes[1])) # O(M) time, O(1) space
    print("Part 2:", part2(pattern, slopes)) # O(M) time and O(1) space

File: src/day3/cpp/solution.cpp
#include <algorithm>
#include <fstream>
#include <string>
#include <tuple>
#include <vector>

#include <fmt/core.h>
#include <fmt/ranges.h>
#include <range/v3/all.hpp>

namespace rs = ranges;
namespace rv = ranges::views;

std::vector<std::string> parse(std::istream &&input) {
  return rs::getlines(input) | rs::to_vector;
}

// non-algorithmic raw-loops solution: good start
auto part1_raw(const std::vector<std::string> &pattern,
               const std::pair<int, int> &slope) {

  auto down = 0, right = 0, trees = 0;
  while (down < pattern.size()) {
    if (pattern[down][right % pattern[down].size()] == '#') {
      trees++;
    }
    down += slope.first;
    right += slope.second;
  }
  return trees;
}

auto part1(std::vector<std::string> pattern, const std::pair<int, int> &slope) {
  const int M = pattern.size(); // iota needs to take ints??
  return rs::count_if(rv::iota(0, M) | rv::stride(slope.first), [&](int row) {
    const auto col = row * slope.second / slope.first;
    return pattern[row][col % pattern[row].size()] == '#';
  });
}

long part2(const std::vector<std::string> &pattern,
           const std::vector<std::pair<int, int>> &slopes) {
  return rs::accumulate(slopes | rv::transform([&](const auto &slope) {
                          return part1(pattern, slope);
                        }),
                        1L, std::multiplies());
}

int main() {
  auto input = parse(std::ifstream{"../../input.txt", std::ios::in});
  const std::vector<std::pair<int, int>> slopes = {
      std::make_pair(1, 1), std::make_pair(1, 3), std::make_pair(1, 5),
      std::make_pair(1, 7), std::make_pair(2, 1)};
  fmt::print("Part 1: {}\n", part1(input, slopes[1]));
  fmt::print("Part 2: {}\n", part2(input, slopes));
  return 0;
}